README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# hershey <img src="man/figures/logo.png" align="right" height=230/>

<!-- badges: start -->

![](http://img.shields.io/badge/cool-useless-green.svg) [![Travis build
status](https://travis-ci.org/coolbutuseless/hershey.svg?branch=master)](https://travis-ci.org/coolbutuseless/hershey)
<!-- badges: end -->

The `hershey` package contains the [Hershey vector
fonts](https://en.wikipedia.org/wiki/Hershey_fonts) in a number of
formats for more convenient use within Rstats.

The Hershey fonts were developed in the 1960s. Each glyph is defined as
collections of staight line segments - hand calculated by Hershey by
sketching them on grid paper\!

This package makes available the coordinates of the stroke endpoints for
every glyph.

You may need this package if:

  - you like retro vector fonts
  - you need a vector font so you can have complete control of the
    output

<img src="man/figures/README-example_header-1.png" width="50%" />

<details closed>

<summary> <span title="Click to open/close" style="color: #496B80;">
Slideshow (click to open/close)</span> </summary>
<img src = "man/figures/fontname/animo.gif">

</details>

<br />

<details closed>

<summary> <span title="Click to open/close" style="color: #496B80;">
Animated Rendering 1 (click to open/close) </span> </summary>
<img src = "man/figures/fontdraw/animo.gif">

</details>

<br />

<details open>

<summary> <span title="Click to open/close" style="color: #496B80;">
Animated Rendering 2 (click to open/close) </span> </summary>
<img src = "man/figures/fontdraw2/animo.gif">

</details>

<br />

## Installation

You can install the development version from
[GitHub](https://github.com/coolbutuseless/hershey) with:

``` r
# install.packages("devtools")
devtools::install_github("coolbutuseless/hershey")
```

## What’s in the box:

  - `hershey_raw`: the original encoding for every glyph (See the
    [vignette](https://coolbutuseless.github.io/package/hershey/articles/hershey-font-format.html)
    for how to parse).
  - `hershey`: a data.frame of the vector strokes for every glyph.
  - `hershey_svg`: a list of SVG paths for every glyph.

Vignettes

  - [Animated stroke
    rendering](https://coolbutuseless.github.io/package/hershey/articles/animated-stroke-rendering.html)
  - [How to generate a sample sheet for a
    font](https://coolbutuseless.github.io/package/hershey/articles/generate-sample-sheet.html)
  - [Description of the raw Hershey font
    format](https://coolbutuseless.github.io/package/hershey/articles/hershey-font-format.html)
  - [Rendering a glyph to
    SVG](https://coolbutuseless.github.io/package/hershey/articles/render-glyph-to-svg.html)
  - [Visualise stroke
    order](https://coolbutuseless.github.io/package/hershey/articles/visualise-stroke-order.html)
  - [3d text
    rendering](https://coolbutuseless.github.io/package/hershey/articles/three-d-text.html)

## Glyph Representations

The following are examples of the representation of the letter `A`
(glyph 34 in the `rowmant` font)

#### `hershey` - First 4 strokes in the data.frame representation

``` r
hershey %>% 
  filter(font == 'rowmant', char == 'A') %>%
  filter(stroke %in% 0:3)
```

    #> # A tibble: 8 x 11
    #>       x     y  left right width stroke   idx glyph font    ascii char 
    #>   <int> <int> <int> <int> <int> <fct>  <int> <int> <chr>   <int> <chr>
    #> 1     0    12   -10    10    20 0          1    34 rowmant    65 A    
    #> 2    -7    -8   -10    10    20 0          2    34 rowmant    65 A    
    #> 3    -1     9   -10    10    20 1          3    34 rowmant    65 A    
    #> 4     5    -9   -10    10    20 1          4    34 rowmant    65 A    
    #> 5     0     9   -10    10    20 2          5    34 rowmant    65 A    
    #> 6     6    -9   -10    10    20 2          6    34 rowmant    65 A    
    #> 7     0    12   -10    10    20 3          7    34 rowmant    65 A    
    #> 8     7    -9   -10    10    20 3          8    34 rowmant    65 A

#### `hershey_raw` - Original encoding

``` r
hershey_raw$rowmant[[34]]
```

    #> [1] " 3001 36H\\RFKZ RQIW[ RRIX[ RRFY[ RMUVU RI[O[ RT[[[ RKZJ[ RKZM[ RWZU[ RWYV[ RXYZ["

#### `hershey_svg` - SVG Path version

``` r
hershey_svg$rowmant[[34]]
```

    #> [1] "M0-12L-7 8M-1-9L5 9M0-9L6 9M0-12L7 9M-5 3L4 3M-9 9L-3 9M2 9L9 9M-7 8L-8 9M-7 8L-5 9M5 8L3 9M5 7L4 9M6 7L8 9"

#### Drawing strokes with `ggplot`

``` r
hershey %>% 
  filter(font == 'rowmant', char == 'A') %>%
  ggplot(aes(x, y, group = stroke)) + 
  geom_path() + 
  geom_point() + 
  coord_equal() + 
  theme_minimal() 
```

<img src="man/figures/README-example_rep-1.png" width="50%" />

## Font Sample Sheets

Font sample sheets show the representation of every glyph in a font.
Click to view the sample sheets for every font in PDF format.

<details closed>

<summary> <span title="Click to open/close" style="color: #496B80;">
Example font: Cursive (click to reveal) </span> </summary>
<img src="man/figures/font/cursive.png">

</details>

<br />

<a href='reference/figures/font/astrology.pdf'>astrology</a> |
<a href='reference/figures/font/cursive.pdf'>cursive</a> |
<a href='reference/figures/font/cursive.png'>cursive</a> |
<a href='reference/figures/font/cyrilc_1.pdf'>cyrilc\_1</a> |
<a href='reference/figures/font/cyrillic.pdf'>cyrillic</a> |
<a href='reference/figures/font/futural.pdf'>futural</a> |
<a href='reference/figures/font/futuram.pdf'>futuram</a> |
<a href='reference/figures/font/gothgbt.pdf'>gothgbt</a> |
<a href='reference/figures/font/gothgrt.pdf'>gothgrt</a> |
<a href='reference/figures/font/gothiceng.pdf'>gothiceng</a> |
<a href='reference/figures/font/gothicger.pdf'>gothicger</a> |
<a href='reference/figures/font/gothicita.pdf'>gothicita</a> |
<a href='reference/figures/font/gothitt.pdf'>gothitt</a> |
<a href='reference/figures/font/greek.pdf'>greek</a> |
<a href='reference/figures/font/greekc.pdf'>greekc</a> |
<a href='reference/figures/font/greeks.pdf'>greeks</a> |
<a href='reference/figures/font/japanese.pdf'>japanese</a> |
<a href='reference/figures/font/japanese2.pdf'>japanese2</a> |
<a href='reference/figures/font/markers.pdf'>markers</a> |
<a href='reference/figures/font/mathlow.pdf'>mathlow</a> |
<a href='reference/figures/font/mathupp.pdf'>mathupp</a> |
<a href='reference/figures/font/meteorology.pdf'>meteorology</a> |
<a href='reference/figures/font/music.pdf'>music</a> |
<a href='reference/figures/font/rowmand.pdf'>rowmand</a> |
<a href='reference/figures/font/rowmans.pdf'>rowmans</a> |
<a href='reference/figures/font/rowmant.pdf'>rowmant</a> |
<a href='reference/figures/font/scriptc.pdf'>scriptc</a> |
<a href='reference/figures/font/scripts.pdf'>scripts</a> |
<a href='reference/figures/font/symbolic.pdf'>symbolic</a> |
<a href='reference/figures/font/timesg.pdf'>timesg</a> |
<a href='reference/figures/font/timesi.pdf'>timesi</a> |
<a href='reference/figures/font/timesib.pdf'>timesib</a> |
<a href='reference/figures/font/timesr.pdf'>timesr</a> |
<a href='reference/figures/font/timesrb.pdf'>timesrb</a>

## Example: All the ’A’s

``` r
glyph_df <- hershey %>% 
  filter(char == 'A') 
  
ggplot(glyph_df) + 
  geom_path(aes(x, y, group = stroke)) + 
  coord_equal() + 
  theme_void() + 
  facet_wrap(~font)
```

<img src="man/figures/README-a-1.png" width="100%" />

## Example: Futura\!

``` r
glyph_df <- hershey %>% 
  filter(font == 'futuram', glyph %in% c(32:46)) 
  
ggplot(glyph_df) + 
  geom_path(aes(x, y, group = stroke)) + 
  coord_equal() + 
  theme_void() + 
  facet_wrap(~glyph, labeller = label_both, ncol = 5)
```

<img src="man/figures/README-futura-1.png" width="100%" />

## Example: Render text string

The function `create_string_df()` will create a data.frame of all the
characters in the given string. It will offset each character by the
known widths of the preceding characters to create a well-spaced set of
points.

``` r
string_df <- hershey::create_string_df(text = "#RStats", font = 'cursive')

ggplot(string_df, aes(x, y, group = interaction(char_idx, stroke))) +
  geom_path() +
  geom_point(size = 1) + 
  coord_equal() +
  theme_void()
```

<img src="man/figures/README-string-1.png" width="100%" />

### References

  - [Wikipedia](https://en.wikipedia.org/wiki/Hershey_fonts)
  - [A great presentation about Hershey and the era of the
    font](http://coopertype.org/event/the_hershey_fonts)
  - [Paul Burke’s description of the
    format](http://paulbourke.net/dataformats/hershey/)
  - A more modern vector font by
    [inconvergent](https://inconvergent.net/):
    [gridfont](https://github.com/inconvergent/gridfont)

### Logo design

The logo is just the letter `H` from the `rowmant` font. Each stroke is
coloured by the stroke number within the glyph and the endpoints are
shown.

## Use Restrictions on Hershey Fonts

**Totally free** (except if you convert them to the proprietary US NTIS
format)

<details closed>

<summary> <span title="Click to open/close" style="color: #496B80;">The
following use restriction was noted in the original font
distribution.</span></summary>

<pre>
    USE RESTRICTION:
    This distribution of the Hershey Fonts may be used by anyone for
    any purpose, commercial or otherwise, providing that:
        1. The following acknowledgements must be distributed with
            the font data:
            - The Hershey Fonts were originally created by Dr.
                A. V. Hershey while working at the U. S.
                National Bureau of Standards.
            - The format of the Font data in this distribution
                was originally created by
                    James Hurt
                    Cognition, Inc.
                    900 Technology Park Drive
                    Billerica, MA 01821
                    (mit-eddie!ci-dandelion!hurt)
        2. The font data in this distribution may be converted into
            any other format *EXCEPT* the format distributed by
            the U.S. NTIS (which organization holds the rights
            to the distribution and use of the font data in that
            particular format). Not that anybody would really
            *want* to use their format... each point is described
            in eight bytes as "xxx yyy:", where xxx and yyy are
            the coordinate values as ASCII numbers.
</pre>

</details>


NEWS.md
# hershey 0.1.0

* Initial release


R/create.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Read a Hershey Font file in .jhf format
#'
#' @param jhf_filename filename
#' @param trim_start = TRUE
#'
#' @return character vector. 1 string per glyph.
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
read_jhf <- function(jhf_filename, trim_start = TRUE) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # One glyph per line.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  jhf <- readLines(jhf_filename)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Find which rows don't start with an integer, and are therefore a
  # continuation of the line above and should be merged with it.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  bad_start <- suppressWarnings(is.na(as.integer(substr(jhf, start = 0, stop = 5))))
  bad_start <- which(bad_start)

  for (ii in rev(bad_start)) {
    jhf[ii - 1] <- paste0(jhf[ii - 1], jhf[ii])
  }

  jhf <- jhf[-bad_start]

  jhf
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a Hershey glyph code to a data.frame
#'
#' A not particularly elegant or fast way to convert a Hershey font code string
#' to a data.frame.
#'
#' @param glyph glyph code character string e.g. "    8  9MWOMOV RUMUV ROQUQ"
#'
#' @return data.frame with x, y and stroke group.
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
convert_glyph_to_df <- function(glyph) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # First 5 chars are the ID (integer)
  # Next 3 chars give the number of vertices as an integer
  # The first 2 points are actual the left + right extents
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  glyph_idx <- substr(glyph, 1, 5)
  Nvertices <- substr(glyph, 6, 8)
  codes     <- substr(glyph, 9, nchar(glyph))
  codes     <- utf8ToInt(codes) - utf8ToInt('R')

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Every pair of characters represents a coordinate pair.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  x <- codes[c(T, F)]
  y <- codes[c(F, T)]

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # First coordinate is actual left/right information
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  left  <- x[ 1]
  x     <- x[-1]
  right <- y[ 1]
  y     <- y[-1]

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # IF there are no strokes, then just make up a 0 point that
  # should probably never get plotted
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  if (length(x) == 0) {
    x <- 0L
    y <- 0L
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Data frame of coordinates.
  # Designate which 'stroke' a point belongs to by looking for " R" i.e. (-50, 0)
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  coords <- data.frame(
    x      = x,
    y      = -y,
    left   = left,
    right  = right,
    width  = right - left,
    stroke = cumsum(x == -50 & y == 0)  # " R" is c(-50, 0)
  )

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Now that 'stroke' is added, we can remove the "pen up" commands
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  coords <- coords[!(coords$x == -50 & coords$y == 0),]

  coords$idx <- seq(nrow(coords))

  coords
}




#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Read a .jhf Hershey font file as a data.frame
#'
#' @param jhf_filename Hershey font file (.jhf)
#'
#' @return data.frame
#'
#' @import tools
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
convert_hershey_font_to_df <- function(jhf_filename) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Read the jhf file in as a character vector. One string per glyph
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  jhf <- read_jhf(jhf_filename)
  font_name <- tools::file_path_sans_ext(basename(jhf_filename))

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Convert each glyph to a data.frame
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  df <- lapply(jhf, convert_glyph_to_df)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Add an index to each glyph
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  df <- mapply(function(x, y) {
    if (!(is.null(x))) {
      x$glyph <- y
    }
    x
  },
  df, seq_along(df),
  SIMPLIFY = FALSE
  )

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Join all glyph data.frames into a font data.frame
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  df <- do.call(rbind, df)
  rownames(df) <- NULL

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Set the font name
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  df$font <- font_name


  df
}














R/string.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create a data.frame of glyph strokes for the given string
#'
#' An error will occur if the string includes ascii which is not part of the font.
#'
#' @param text text string
#' @param font font name. default 'rowmant'
#'
#' @return data.frame with coorindates for all glyphs with characters
#' offset appropriately. 'char_idx' is the index of the character within the
#' given 'text' string.
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
create_string_df <- function(text, font = 'rowmant') {

  hershey <- hershey::hershey

  if (!font %in% hershey$font) {
    stop("create_string_df(): No such font: ", font, call. = FALSE)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Select the requested font.
  # Set an ordering so stuff doesn't get messed up during the merge
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  font_df <- hershey[hershey$font == font, ]

  ascii  <- utf8ToInt(text)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Check the font has the required characters
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  missing <- unique(ascii[!ascii %in% font_df$ascii])
  if (length(missing) > 0) {
    stop("create_string_df(): Requested ascii characters that don't exist in font: ",
         font, " ", deparse(intToUtf8(missing)), call. = FALSE)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Create a data.frame of all the requested characters and their index within
  # the initial text
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  text_df <- data.frame(
    ascii    = ascii,
    char_idx = seq_along(ascii)
  )

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Create a dataset with all the requested characters and the glyphs which
  # generate them
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  string_df <- merge(text_df, font_df, sort=FALSE)
  string_df <- string_df[with(string_df, order(char_idx, stroke, idx)), ]

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Need to work out how far to offset characters after the first character
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  offsets_df <- unique(string_df[,c('char_idx', 'char', 'left', 'right')])

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # The difference between centres of two characters is the sum of the right boundary of
  # first and the left_boundary of the second character
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  kern <- with(offsets_df, head(right, -1) - tail(left, -1))
  kern <- c(0, kern)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Cumulatively offset all characters
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  offsets_df$offset <- cumsum(kern)
  string_df        <- merge(string_df, offsets_df, sort=FALSE)
  string_df$x      <- string_df$x + string_df$offset

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Merging totally borks the order, so put everything back in place
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  string_df   <- string_df[with(string_df, order(char_idx, stroke, idx)), ]

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # add an overall stroke index
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  string_df$stroke_idx <- as.integer(interaction(string_df$stroke, string_df$char_idx))

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Make it a tibble and return
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  class(string_df) <- c('tbl_df', 'tbl', 'data.frame')
  string_df
}



if (FALSE) {

  text = "Great White Shark"
  font = 'timesrb'

  library(ggplot2)
  string_df <- create_string_df(text = text, font = font)

  ggplot(string_df) +
    geom_path(aes(x, y, group = interaction(char_idx, stroke))) +
    coord_equal() +
    theme_minimal()


}


R/hershey.R


#' Hershey vector fonts as a single data.frame
#'
#' @format data.frame of all hershey vector fonts
#' \describe{
#'   \item{x,y}{coordinates of endpoints of a single line segment. A stroke consistes of multiple line segements. A glyph consistes of multiple strokes.}
#'   \item{left,right}{Plotting extents of individual glyph}
#'   \item{width}{Width of individual glyph}
#'   \item{stroke}{Index of the stroke within this glyph}
#'   \item{idx}{Index of the point within this glyph}
#'   \item{glyph}{Index of the glyph within this font}
#'   \item{font}{Font name}
#'   \item{ascii}{ASCII code. Set to -1 if it doesn't match a standard ASCII character}
#'   \item{char}{ASCII character representation e.g. 'a'.  Set to empty string if doesn't match a standard ASCII character}
#' }
"hershey"


#' Hershey vector fonts as a list of character strings of the original glyph encodings
#'
#' See vignette for a description of this format.
"hershey_raw"

#' Hershey vector fonts as a list of character strings of SVG paths for each glyph
"hershey_svg"



data-raw/prepare-hershey-df.R



suppressPackageStartupMessages({
  library(dplyr)
  library(purrr)
  library(ggplot2)
})


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Read in all hershey font as data.frame
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jhf_filenames <- list.files(here::here("data-raw/hershey"), pattern = '*.jhf', full.names = TRUE)
hershey <- jhf_filenames %>%
  map(convert_hershey_font_to_df) %>%
  bind_rows() %>%
  as.tbl() %>%
  mutate(stroke = as.factor(stroke))


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Join in ascii mapping
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ascii <- data.frame(
  ascii = 32:126,
  char  = strsplit(intToUtf8(32:126), '')[[1]],
  stringsAsFactors = FALSE
) %>%
  mutate(glyph = seq(n()))

hershey <- hershey %>% left_join(ascii, by = 'glyph')



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# But then there are a lot of fonts that we don't want ascii mapping
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ii <- hershey$font %in% c('markers', 'japanese', 'japanese2', 'symbolic')
hershey$char [ii] <- ''
hershey$ascii[ii] <- -1L



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# NA out individual glyphs in individual fonts
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
good <- c(1, 3, 9, 10, 13, 15, 17:26, 34:59, 66:91)
ii <- hershey$font == 'astrology' & !hershey$glyph %in% good
hershey$char [ii] <- ''
hershey$ascii[ii] <- -1L


good <- c(1, 6, 9, 10, 12:26, 29:31, 34:62, 92, 94)
ii <- hershey$font %in% c('mathlow', 'mathupp') & !hershey$glyph %in% good
hershey$char [ii] <- ''
hershey$ascii[ii] <- -1L


good <- c(1, 6, 9, 10, 11, 13:26, 32:62, 66:92, 94)
ii <- hershey$font == 'meterology' & !hershey$glyph %in% good
hershey$char [ii] <- ''
hershey$ascii[ii] <- -1L


good <- c(1, 2, 17:27, 34:62, 64:95)
ii <- hershey$font == 'music' & !hershey$glyph %in% good
hershey$char [ii] <- ''
hershey$ascii[ii] <- -1L





#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Replace NAs with zero-width blank as it looks nicer :)
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hershey <- hershey %>%
  tidyr::replace_na(list(char = '', ascii = -1L))


usethis::use_data(hershey, overwrite = TRUE)




#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Read in hershey fonts as glyph encodings
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hershey_raw <- jhf_filenames %>%
  map(read_jhf) %>%
  set_names(tools::file_path_sans_ext(basename(jhf_filenames)))


usethis::use_data(hershey_raw, overwrite = TRUE)



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Convert glyphs to SVG
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
convert_stroke_to_path <- function(stroke_df) {
  pairs      <- paste(stroke_df$x, -stroke_df$y)
  line_to    <- paste0('L', pairs)
  line_to[1] <- sub('L', 'M', line_to[1])
  path       <- paste(line_to, collapse = " ")
  path
}


convert_glyph_to_path <- function(glyph_df) {
  strokes       <- split(glyph_df, as.integer(glyph_df$stroke))
  path_segments <- vapply(strokes, convert_stroke_to_path, character(1))
  path          <- paste(path_segments, collapse = "")
  path <- gsub(" L", "L", path)
  path <- gsub(" -", "-", path)
  path
}

convert_font_to_paths <- function(font_df) {
  glyphs <- split(font_df, font_df$glyph)
  vapply(glyphs, convert_glyph_to_path, character(1))
}



hershey_svg <- split(hershey, hershey$font) %>%
  map(convert_font_to_paths)



usethis::use_data(hershey_svg, overwrite = TRUE)

if (FALSE) {
  svg <- glue::glue('<svg width="200" height="300" viewBox="-10 -10 20 20" xmlns="http://www.w3.org/2000/svg">
  <path d="{hershey_svg$cursive[[36]]}" stroke="black" fill="transparent" stroke-width="0.5">
             <animateTransform attributeName="transform"
                          attributeType="XML"
                          type="rotate"
                          from="0"
                          to="360"
                          dur="10s"
                          repeatCount="indefinite"/>
  </path>
</svg>')
  writeLines(svg, "test.svg")
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create the pdf for the given a data.frame with a single font
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
create_font_pdf <- function(font_df) {

  stopifnot(length(unique(font_df$font)) == 1L)
  font_name <- font_df$font[1]

  font_df <- font_df %>% mutate(
    label = sprintf("%03i %s", glyph, char)
  )

  p <- ggplot(font_df) +
    geom_path(aes(x, y, group=stroke)) +
    theme_void() +
    coord_equal() +
    theme(legend.position = 'none') +
    facet_wrap(~label) +
    labs(title = font_name)

  pdf_name <- paste0(font_name, ".pdf")
  pdf_name <- here::here("man", "figures", "font", pdf_name)

  ggsave(pdf_name, plot = p, width = 15, height = 15)
}



if (FALSE) {
  font_df <- hershey %>% filter(font == 'astrology')
  create_font_pdf(font_df)
} else {
  hershey %>%
    group_split(font) %>%
    walk(create_font_pdf)
}

